src/lib.rs

```diff
@@ -1021,6 +1045,339 @@ safety_comment! {
+        ///
+        /// assert_eq!(packet.length, 4);
+        /// assert_eq!(packet.body, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
+        ///
+        /// {
+        ///     let (packet, rest) = packet.split_at_mut(packet.length as usize).unwrap();
+        ///     assert_eq!(packet.length, 4);
+        ///     assert_eq!(packet.body, [1, 2, 3, 4]);
+        ///     assert_eq!(rest, [5, 6, 7, 8, 9]);
+        ///
+        ///     rest.fill(0);
+        /// }
+        ///
+        /// assert_eq!(packet.length, 4);
+        /// assert_eq!(packet.body, [1, 2, 3, 4, 0, 0, 0, 0, 0]);
+        /// ```
+        #[inline]
+        fn split_at_mut(&mut self, l_len: usize) -> Option<(&mut Self, &mut [Self::Elem])> {
+            match MetadataOf::new_in_bounds(self, l_len) {
+                Some(l_len) if l_len.padding_needed_for() == 0 => {
+                    // SAFETY: We have ensured both that:
+                    // 0. `l_len <= self.len()` (by post-condition on
+                    //    `MetadataOf::new_in_bounds`)
+                    // 1. `l_len.padding_needed_for() == 0` (by guard on match arm)
+                    Some(unsafe { self.split_at_mut_unchecked(l_len.get()) })
+                }
+                _ => None,
+            }
+        }
+    }
+
+    // SAFETY: `[T]`'s trailing slice is `[T]`, which is trivially aligned.
+    unsafe impl<T> SplitAt for [T] {
+        type Elem = T;
+
+        #[inline]
+        fn only_derive_is_allowed_to_implement_this_trait()
+        where
+            Self: Sized,
+        {
+        }
+    }
+
+    /// Splits `T` in two.
+    ///
+    /// # Safety
+    ///
+    /// The caller promises that:
+    /// 0. `l_len.get()` is not greater than the length of `ptr`'s trailing
+    ///    slice.
+    /// 1. if `I::Aliasing` is [`Exclusive`] or `T` permits interior mutation,
+    ///    then `l_len.padding_needed_for() == 0`.
+    #[inline(always)]
+    unsafe fn ptr_split_at_unchecked<'a, T, I, R>(
+        ptr: Ptr<'a, T, I>,
+        l_len: MetadataOf<T>,
+    ) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+    where
+        I: invariant::Invariants,
+        T: ?Sized + pointer::Read<I::Aliasing, R> + SplitAt,
+    {
+        let inner = ptr.as_inner();
+
+        // SAFETY: The caller promises that `l_len.get()` is not greater than
+        // the length of `self`'s trailing slice.
+        let (left, right) = unsafe { inner.split_at_unchecked(l_len) };
+
+        // Lemma 0: `left` and `right` conform to the aliasing invariant
+        // `I::Aliasing`. Proof: If `I::Aliasing` is `Exclusive`, the caller
+        // promises that `l_len.padding_needed_for() == 0`. Consequently, by
+        // post-condition on `PtrInner::split_at_unchecked`, there is no
+        // trailing padding after `left`'s final element that would overlap into
+        // `right`. If `I::Aliasing` is shared, then overlap between their
+        // referents is permissible.
```

Comment:
Mention interior mutation in both cases (ie, "If `I::Aliasing` is `Exclusive` or `T` permits interior mutation," and "If `I::Aliasing` is `Shared` and `T` forbids interior mutation").

Reply:
https://github.com/google/zerocopy/compare/4aecf267464e9c919342e86664ec339daba7a2b4..788d4ccaec59e3ec0e931131da4bdef35e611d2b#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R1301-R1307